addons/mail/models/stage_time_mixin.py

```diff
@@ -0,0 +1,111 @@
+                record.time_per_stage = record.time_per_stage
+            else:
+                record.time_per_stage = record._get_json_ms_in_stages()
+
+    def _get_json_ms_in_stages(self):
+        """
+        :return: string: JSON {stage_id (str): ms (float)}
+
+        e.g.:    '{"3": 1000, "2": 44211.3331}'
+        """
+
+        stages_with_time_json = "{}"
+        # Should not compute if the record was just created or created without stage and then assigned a stage
+        if (self.env.cr.precommit.data
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}']
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id] is not None
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field]
+                and self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field] != self.stage_id):
+            mail_tracking_records = self.env['mail.tracking.value'].search([
+                ('mail_message_id.res_id', '=', self.id),
+                ('mail_message_id.model', '=', self._name),
+                ('field', '=', self.env.ref(f'{self._original_module}.field_{self._table}__{self._stage_field}').id)
+            ], order='create_date desc')
+
+            tracking_records_with_order = self._get_stage_moves_from_tracking(mail_tracking_records)
+            time_in_stage = self._get_time_spent_in_stage(tracking_records_with_order)
+            if time_in_stage:
+                stages_with_time_json = self._json_stage_id_with_time(time_in_stage)
+        return stages_with_time_json
+
+    def _get_stage_moves_from_tracking(self, mail_tracking_value_records):
+        stage_moves = []
+        stages_order = self._get_stages_order()
+
+        # Because there is no tracking message for this stage change yet
+        stage_moves.append({
+            'old_stage_id': self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field].id,
+            'stage_move_date': fields.datetime.now(),
+            'old_order': stages_order[self.env.cr.precommit.data[f'mail.tracking.{self._name}'][self.id][self._stage_field].id],
+            'new_order': stages_order[self[self._stage_field].id],
+        })
+
+        for record in mail_tracking_value_records:
+            if record.old_value_integer in stages_order and record.new_value_integer in stages_order:
+                stage_moves.append({
+                    'old_stage_id': record.old_value_integer,
+                    'stage_move_date': record.create_date,
+                    'old_order': stages_order[record.old_value_integer],
+                    'new_order': stages_order[record.new_value_integer],
+                })
+        return stage_moves
+
+    def _get_stages_order(self):
+        stages_order = {}
+        order = 0
+
+        for stage in self.env[self._fields[self._stage_field]._related_comodel_name].search([]):
```

Comment:
Good point. Do you have an idea on how to proceed?
Right now, I'm trying to get the field domain from a string `self._fields[self._stage_field].domain` which gives something like `"['|', ('team_id', '=', False), ('team_id', '=', team_id)]"`, and evaluating it with `safe_eval()`.
However, `safe_eval()` cannot evaluate `team_id` as it would need to be `self.team_id`
Is there any way to get the domain already evaluated, like `"['|', ('team_id', '=', False), ('team_id', '=', 1)]"` ?
Thanks!

Reply:
What comes out of the discussion is that the `domain` attribute of the `stage_id` many2one field is only a front-end filter. Which means a record can potentially be assigned to any stage, in which case we need all stages for our computations

Reply:
Indeed :)

This should be fine, there doesn't seem to be any other indicator that a stage is not shared, and evaluating in the back-end seems error-prone at best.

The query should remain relatively fast even for a few hundred stages and we don't iterate over it later.
Maybe we can scratch off a little perf by using iterators and comprehension dicts instead.
(also makes for fewer lines)

```
stages = self.env[self._fields[self._stage_field]._related_comodel_name].search([])
return {stage: order for order, stage in enumerate(stage)}
```